src/ansi-c/c_typecheck_type.cpp

```diff
@@ -948,6 +948,41 @@ void c_typecheck_baset::typecheck_compound_body(
 
         typecheck_type(new_component.type());
 
+        // the rules for anonymous members depend on the type and the compiler,
+        // just bit fields are accepted everywhere
+        if(
+          new_component.get_name().empty() &&
+          new_component.type().id() != ID_c_bit_field)
+        {
+          if(config.ansi_c.mode == configt::ansi_ct::flavourt::VISUAL_STUDIO)
+          {
+            // Visual Studio rejects anything other than a struct or union
+            // declaration, but also accepts those when they introduce a tag
+            if(
+              new_component.type().id() != ID_struct_tag &&
+              new_component.type().id() != ID_union_tag)
+            {
+              error().source_location = source_location;
+              error() << "no members defined" << eom;
+              throw 0;
+            }
+          }
+          else
+          {
+            // GCC and Clang ignore anything other than an untagged struct or
+            // union; we could print a warning, but there isn't any ambiguity in
+            // semantics here. Printing a warning could elevate this to an error
+            // when compiling code with goto-cc with -Werror.
+            if(
+              (new_component.type().id() != ID_struct_tag &&
+               new_component.type().id() != ID_union_tag) ||
+              follow(new_component.type()).find(ID_tag).is_not_nil())
```

Comment:
What does the follow do here?

Reply:
This is to distinguish untagged (anonymous) structs/unions from ones that are named.

Reply:
I have extended the comment above this code now.

Reply:
May I suggest using `to_tag_type` to make this clear.

Reply:
I've rewritten it to use `to_struct_tag_type` and `to_union_tag_type`.